src/hooks/context/useTransactionsStatus.tsx

```diff
@@ -0,0 +1,115 @@
+import { useApiCallsContext } from '../../contexts/ApiCallsContextProvider';
+import useLaneId from '../chain/useLaneId';
+import { useEffect } from 'react';
+import { TransactionActionCreators } from '../../actions/transactionActions';
+import isEqual from 'lodash/isEqual';
+import { MessageActionsCreators } from '../../actions/messageActions';
+import usePrevious from '../react/usePrevious';
+
+export default function useTransactionsStatus(
+  transactions: TransactionStatusType[],
+  evaluatingTransactions: boolean,
+  dispatchTransaction: Function
+) {
+  const { dispatchMessage } = useUpdateMessageContext();
+  const { getValuesByChain } = useChainGetters();
+
+  const subscriptions = useSubscriptionsContext();
+  const prevSubscriptions = usePrevious(subscriptions);
+
+  const {
+    sourceChainDetails: { chain: currentSourceChain }
+  } = useSourceTarget();
+  const apiCalls = useApiCallsContext();
+  const laneId = useLaneId();
+
+  useEffect(() => {
+    const getTransactionStatus = async () => {
+      dispatchTransaction(TransactionActionCreators.setEvaluatingTransactionsStatus(true));
+      const updatedTransactions = await Promise.all(
+        transactions.map(async (transaction: TransactionStatusType) => {
+          if (transaction.status === TransactionStatusEnum.COMPLETED) {
+            return transaction;
+          }
+
+          const { sourceChain, targetChain } = transaction;
+          const { api: targetApi } = getValuesByChain(targetChain);
+          const { sourceRole, targetRole } = getChainSubscriptionsKey({
+            currentSourceChain,
+            sourceChain
+          });
+
+          const sourceSubscriptions = subscriptions[sourceRole];
+          const targetSubscriptions = subscriptions[targetRole];
+
+          const updatedTransaction = await handleTransactionUpdates({
+            transaction,
+            sourceSubscriptions,
+            targetSubscriptions,
+            apiCalls: {
+              targetApi,
+              ...apiCalls
+            },
+            laneId,
+            dispatchMessage
+          });
+
+          return updatedTransaction;
+        })
+      );
+
+      if (!isEqual(transactions, updatedTransactions)) {
+        dispatchTransaction(TransactionActionCreators.updateTransactionsStatus(updatedTransactions));
+      }
+
+      dispatchTransaction(TransactionActionCreators.setEvaluatingTransactionsStatus(false));
+    };
+    const transactionsInProgress = transactions.find(({ status }) => status === TransactionStatusEnum.IN_PROGRESS);
+    const updatedSubscriptions = prevSubscriptions !== subscriptions;
+
+    if (!evaluatingTransactions && transactionsInProgress && updatedSubscriptions) {
+      try {
+        getTransactionStatus();
+      } catch (e) {
+        dispatchMessage(MessageActionsCreators.triggerErrorMessage({ message: e }));
+        dispatchTransaction(TransactionActionCreators.setEvaluatingTransactionsStatus(false));
+      }
```

Comment:
IMHO that would work really nicely, cause you separate the logic of `getTransactionStatus` from enacting the values to the reducer - i.e. there is a single co-located place where the updates happen, instead of having the scattered.

```suggestion
     dispatchTransaction(TransactionActionCreators.setEvaluatingTransactionsStatus(true));
     getTransactionStatus()
       .then(updatedTransactions => {
         if (!isEqual(transactions, updatedTransactions)) {
           dispatchTransaction(TransactionActionCreators.updateTransactionsStatus(updatedTransactions));
         }
         dispatchTransaction(TransactionActionCreators.setEvaluatingTransactionsStatus(false));
       })
       .catch( e => {
         dispatchMessage(MessageActionsCreators.triggerErrorMessage({ message: e }));
         dispatchTransaction(TransactionActionCreators.setEvaluatingTransactionsStatus(false));
       });
```

Reply:
i really like this approach. I will definitely implement it.

Reply:
@tomusdrw , Finally i didn't implement your proposal because i used the the new version of `genericCall` for managing the async call and reducer state management in a consistent and tested way. 
I think that closely we are reaching the final model.